### PR TITLE
Predicate on our "well-known attributes"

### PR DESCRIPTION
**Pull request by alexsnaps**
fix to #112

### CODE REVIEW

**Comment by alexsnaps** [comment]
This may need to supersede #110 ... stil wip tho

**Comment by alexsnaps** [comment]
[tl;dr](https://github.com/Kuadrant/wasm-shim/pull/113/commits/5c558d4519a48bde4af8da179dcd9701d7e167d5#diff-f6344880d0def0b0b03411b5e9439442d8713a0447ad01d258ec087595dbc7e9R256-R260) this replaces `PatternExpression` with CEL `Predicates`, while keeping the former as a working fallback... 

Of interest if probably [this change](https://github.com/Kuadrant/wasm-shim/pull/113/commits/bf4f2f2754ab4460da86f5a7a59d627d9420c42b) as well, where the whole `fn` needs to be replace as `if cfg!(test)` within the method would work, but won't make the linker happy on anything but wasm...

**Comment by eguzki** [comment]
I would change the PR tittle. That goes on the auto-generated release notes.

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:138]: >  the value is the string "true"

What is the problem  there? `auth.identity.anonymous` returns a string. We should document that and the user should deal with that. 

Besides, this something we can change from authorino, can't we? I mean, `auth.identity.anonymous` could be `bool` not string.

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:154]: where are those `destination` and `port` coming from?

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:133]: `wasm.kuadrant.{key}` or `kuadrant.{key}` ?

File: src/configuration.rs
Comment (eguzki) [inline src/configuration.rs:171]: I do not get this comment. For backwards compat?

why not :fire: to `PatternExpressions`?

File: src/configuration/action.rs
Comment (eguzki) [inline src/configuration/action.rs:30]: we need to document clearly that `predicates` take precedence and when at least one predicate is being defined, conditions are not even being evaluated.

**Review by eguzki — COMMENTED** [review]
review still WIP

File: src/data/attribute.rs
Comment (alexsnaps) [reply src/data/attribute.rs:154]: ... and json numbers are floats: `1` in JSON is `1_f64` using rust notation. It is not an integer.

File: src/data/attribute.rs
Comment (alexsnaps) [reply src/data/attribute.rs:133]: `wasm.` prefix is automatically added by envoy. So you store `foobar`, but to retrieve it back you read `wasm.foobar`...

File: src/configuration.rs
Comment (alexsnaps) [reply src/configuration.rs:171]: Yes! We should burn them as soon as the kuardrant operator doesn't use them to configure this wasm filter anymore...

File: src/configuration/action.rs
Comment (alexsnaps) [reply src/configuration/action.rs:30]: We can... if this does NOT 🔥 by the release.

**Comment by alexsnaps** [comment]
Just to be clear about the state of this... now that #110 was merged, while broken, let's not merge this quite yet. 
I took @didierofrivia work over in supporting "not so well known attribute", i.e. the data coming from authorino. I'll do that in 2 passes, scalar values (e.g. `auth.identity.anonymous` being resolved as being of type `bool` and then supporting any `expression` coming from authorino, expressed as JSON, in there too: List & Object)
I should get the former done by EOD EST, possibly the latter too...

File: src/configuration.rs
Comment (eguzki) [reply src/configuration.rs:171]: yeah, that would allow us to merge this without waiting for operator to implement that.

**Comment by alexsnaps** [comment]
[this change](https://github.com/Kuadrant/wasm-shim/pull/113/commits/f6ca3f6ecfb7492b7102aea744990cf6c12818b3) has the first step, but I think the latter step is _not_ required because of how the `struct` is decomposed in [process_metadata](https://github.com/Kuadrant/wasm-shim/blob/f8690551744a4edfd53b3b71fb5242b0cab8027b/src/data/attribute.rs#L173-L190).

File: src/data/cel.rs
Comment (eguzki) [inline src/data/cel.rs:29]: When a expression contains `request.headers.X-MY-HEADER`, that path is not a known attribute. Hence, the attribute is read expecting a String and returned bytes are parsed as String. 

However, the prefix of it, i.e. `request.headers`, is well known, a `map<string, string>`, how could we leverage that to be smarter?

**Review by eguzki — CHANGES_REQUESTED** [review]
nice peace of work!

Some minor comments dropped.

Doc is required

File: src/data/cel.rs
Comment (alexsnaps) [reply src/data/cel.rs:29]: Yes... tho probably for now I would read the entire map and let the user deal as such with it. 
Another way is to have `properties` follow maps, as we know we can ask envoy to resolve these for us. 
Cause technically, that's the [way](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/main/src/hostcalls.rs#L140-L164)

File: src/data/cel.rs
Comment (alexsnaps) [reply src/data/cel.rs:345]: Yes, right now that is NOT supported. 
Ideally the 2nd entry would need to be filtered out. So I was thinking that one day we could sort, so that the shortest path are always first, and subsequent ones can "just" be ignored. 
It's not an issue right now, as you can't do anything with `auth` for instance right now really. Not entirely true, but say nothing useful. Also, that is also always better expressed without requiring the whole shorter path to be resolved. One exception is `request.headers` but that needs special attention anyways.

File: src/data/cel.rs
Comment (alexsnaps) [reply src/data/cel.rs:84]: Well yes, but decided _not_ because it has the concept of _known attribute types_ that are CEL. So it could actually be `CelTypedAttribute` instead...

File: src/data/cel.rs
Comment (eguzki) [reply src/data/cel.rs:345]: Have you considered building another type of tree structure?

```rust
struct Node {
     value: Option<Attribute>,  
     children: Vec<Node>,
} 

struct Tree {
     roots: Vec<Node>,
}
```

Comment (alexsnaps) [reply src/data/cel.rs:345]: To what purpose? I assume you're talking about this input:

```
[
   Attribute(["auth"]),
   Attribute(["auth", "identity")
]
```
yes? If so, let's zoom back for a second. if you access `auth`, I need to hydrate `auth` in full... So that `["auth", "identity"]` is of no use anymore, at the `identity` member was hydrated as part of `auth`, right? 
Or am I missing the use case you are trying to address here?

Comment (alexsnaps) [reply src/data/cel.rs]: I've ordered and skip an "more precise" attribute when a Value was resolved earlier in the graph.
So if `auth` was resolved, no need to resolve `auth.identity`, as its a member (hopefully) of `auth` that's now fully hydrated, see [here](https://github.com/Kuadrant/wasm-shim/pull/113/commits/dbdf7e10a7f3f9286143d3ad4ad655255f70e9fc)

File: src/data/cel.rs
Comment (alexsnaps) [reply src/data/cel.rs:29]: Also, this `request.headers.X-MY-HEADER` is not a valid expression, it should be `request.headers["X-MY-HEADER"]`

File: src/data/cel.rs
Comment (eguzki) [reply src/data/cel.rs]: I was thinking on 

```
contains(request.headers, "X-MY-HEADER") && request.headers.OTHER-HEADER == "some-value"
```

But, no need to address this kind of issues now. Nitpicking here.

**Comment by alexsnaps** [comment]
Alright, it should "all" work. i.e. now we store JSON representation of the `auth` data coming back from authorino, so to not lose the type information... We even get integers! I also added support for getting to headers...

File: src/data/property.rs
Comment (didierofrivia) [inline src/data/property.rs:49]: I like this way of defining the fn when it's meant to be mocked for testing and not. Might be worthy to follow the same pattern for the operation dispatcher and grpc message req/res //TODO:(didierofrivia)

**Review by didierofrivia — APPROVED** [review]
LGTM. 💪🏼 🥇 🙏🏼

File: src/data/cel.rs
Comment (alexsnaps) [reply src/data/cel.rs:86]: Which btw is why it's in the `cel` ~file~ module.

File: src/configuration/action.rs
Comment (alexsnaps) [inline src/configuration/action.rs:63]: This is how we've send data to limitador so far, losing the type information... just a string: `true`, `42`, `bob`

File: src/configuration/action.rs
Comment (alexsnaps) [inline src/configuration/action.rs:70]: Wondering if we should skip that descriptor entry... this _could_ happen if a value is explicitly set to `null` in `auth` ...

**Comment by alexsnaps** [comment]
Status update: this 🔥 the need for anything but CEL. While the previous ways are still there, they should _NOT_ be used anymore and comes with bunch of caveat... I'll let the code speak for itself as how one can shoot themselves in the foot by going to raw properties/attribute values.

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:132]: I was thinking that the `auth` prefix should come from `pub fn process_auth_grpc_response` method.

Then, `pub fn store_metadata(metastruct: &Struct)` could be `pub fn store_metadata(prefix: &str, metastruct: &Struct)`

My diff

```diff
diff --git a/src/data/attribute.rs b/src/data/attribute.rs
index 9e65caa..7888462 100644
--- a/src/data/attribute.rs
+++ b/src/data/attribute.rs
@@ -126,10 +126,10 @@ pub fn set_attribute(attr: &str, value: &[u8]) {
     };
 }
 
-pub fn store_metadata(metastruct: &Struct) {
+pub fn store_metadata(prefix: &str, metastruct: &Struct) {
     let metadata = process_metadata(metastruct, String::new());
     for (key, value) in metadata {
-        let attr = format!("{KUADRANT_NAMESPACE}\\.auth\\.{key}");
+        let attr = format!("{KUADRANT_NAMESPACE}\\.{prefix}\\.{key}");
         debug!("set_attribute: {attr} = {value}");
         set_attribute(attr.as_str(), value.into_bytes().as_slice());
     }
diff --git a/src/service/auth.rs b/src/service/auth.rs
index c8e4a4d..8364a8b 100644
--- a/src/service/auth.rs
+++ b/src/service/auth.rs
@@ -128,7 +128,7 @@ impl AuthService {
     ) -> Result<(), StatusCode> {
         if let GrpcMessageResponse::Auth(check_response) = auth_resp {
             // store dynamic metadata in filter state
-            store_metadata(check_response.get_dynamic_metadata());
+            store_metadata("auth", check_response.get_dynamic_metadata());
 
             match check_response.http_response {
                 Some(CheckResponse_oneof_http_response::ok_response(ok_response)) => {
```

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:133]: What about this `debug!` macro in `pub fn set_attribute(attr: &str, value: &[u8])` ?

File: src/data/attribute.rs
Comment (eguzki) [inline src/data/attribute.rs:132]: Can you please add a comment saying that attributes set here are accessible with `wasm.*` prefix?

File: src/configuration.rs
Comment (eguzki) [inline src/configuration.rs:28]: can we make `key` optional? it would default to the expression value (not the referenced value, though).

**Review by eguzki — COMMENTED** [review]
Left some comments and food for thoughts. 

This is not about requesting changes. 

From my side, this is ready to be merged

File: src/configuration.rs
Comment (alexsnaps) [reply src/configuration.rs:28]: I think it makes for a very horrible default, but... 🤷

**Comment by alexsnaps** [comment]
@eguzki This is my gift to you... you all of course :)
Do whatever you want with the PR! You have my blessing. I need to move on here...

**Review by eguzki — APPROVED** [review]

**Pull request merged**
